### PR TITLE
[Popper] Use `floating-ui`

### LINKED ISSUE #370: [Popper] Improve calculations to account for parent positioning / scroll offsets
This is for cases when people aren't rendering in portals.

See https://github.com/radix-ui/primitives/issues/367 for example.

### LINKED ISSUE #557: [Popover] Better collision prevention
## Feature request

When using `align="start"` or `align="end"` the alignment may flip abruptly when a collision happens. Sometimes this introduces just another collision, which neither uses the available space efficiently, nor keeps the preferred alignment intact.

For example, when using `align="start" side="right"` the desired behaviour would be:
- If `align="start"` results in a collision, try `align="center"`
- If `align="center"` does not resolve the collision, try `align="end"`
- If `align="end"` does not resolve the collision, revert to `align="start"`

Current behaviour:

https://user-images.githubusercontent.com/8441036/111156630-a5ec8400-859e-11eb-906a-3fc953ce4a10.mov

### LINKED ISSUE #740: [Popover] content positioned incorrectly on Safari when combined with Framer Motion scale
## Bug report
When using the Popover component with it's Popover.Content set to `align="end"` and animating its scale with Framer Motion it's positioned incorrectly on the x-axis on Safari. 

### Current Behavior
The popover content is offset to the right.

Screenshot 2021-06-22 at 20 47 58

### Expected behavior
The popover content should be aligned to the trigger. This works in Chrome/FF.

### Reproducible example

https://codesandbox.io/s/radix-framer-bug-myc5h

### Additional context

Very specific bug for my use-case ( opening a right-aligned popover and animating it's scale with framer motion), and not sure if fixable but it works as expected in Chrome/FF

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) |    @radix-ui/react-popover     |    0.0.18     |
| React            | n/a     |    17.0.2     |
| Browser          |    Safari     |     14.1    |
| Assistive tech   |         |         |
| Node             | n/a     |    16.0.0     |
| npm/yarn         |    yarn     |     1.22.10    |
| Operating System |    MacOS     |    11.3.1     |

### LINKED ISSUE #781: [Popper] Issue when trigger is position fixed?
### Current Behavior

CleanShot 2021-07-13 at 15 22 59

### LINKED ISSUE #800: [Popover] Off-center arrow position with scale transform in Safari
### Current Behavior

In Safari, using a scale transform to animate `Popover.Content` makes the arrow element go off-center.

Screenshot 2021-07-27 at 1 04 39 PM

P.S. Using code from Radix documentation: https://www.radix-ui.com/docs/primitives/components/popover#origin-aware-animations

### Expected behavior

Correct positioning of the arrow, and content elements.

### Reproducible example

CodeSandbox: https://codesandbox.io/s/radix-popover-animation-2w1p0?file=/src/App.js

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) | @radix-ui/react-popover | 0.0.19 |
| Stitches             | n/a | 0.2.3 |
| React            | n/a     | 17.0.2 |
| Browser          | Safari | 15.0 (17612.1.22.1.4) |
| Assistive tech   | - | - |
| Node             | n/a     | 16.5.0 |
| npm/yarn         | npm | 7.19.1 |
| Operating System | macOS | 12.0 Beta |

### LINKED ISSUE #1445: Component `Hover Card` does not respect inner containers
### Current Behavior

At the moment component `Hover Card` is working only based on the viewport and there are some edge cases where it stops working if we're trying to use it inside inner scrollable containers.

<details>
<summary>Demo 👀</summary>
<video src="https://user-images.githubusercontent.com/9263565/171867455-365af036-c3c8-4e1a-aff1-75077301f6d5.mov" />
</details>

I found a temporary solution, however it does not solve the issue itself:

1. Add `portalled={false}` to `HoverCard.Content` component.
2. Add `z-index: 0` to `HoverCard.Content` component.
3. Add `background-color` as well as `z-index: 1` to containers above and below.

<details>
<summary>Demo 👀</summary>
<video src="https://user-images.githubusercontent.com/9263565/171867948-bf72793c-5088-47f2-b784-bf2a581bfa42.mov" />
</details>

As you can see it solves the issue, but this approach leads to another bug. We're losing `avoidCollisions` and `collisionTolerance` features. The card itself is being hidden, however it does not avoid collisions anymore and being invisible, the result is bad UX.

### Expected behavior

`HoverCard` component should respect inner container boundaries and try to avoid collisions.

### Suggested solution

I guess there are few solutions and I would be more than happy to contribute and fix it by myself, but I am not sure which solution is the best.

1. Hide `HoverCard.Content` component on scroll. I am not sure about this one, because we should make sure that we're not scrolling for example inside `HoverCard.Content` component. Also I am not sure if it should be working like that by default, maybe we should just simply add additional prop?
2. Add ability to pass the reference of parent container, in this case, `HoverCard.Content` component would respect container boundaries instead of viewport. I'm not sure if this solution can be easily implemented.
3. Your suggestions? 💡

### LINKED ISSUE #1464: [DropdownMenu] DropdownMenu doesn't locate properly.
### Current Behavior
I'm using dropdown menu for my i18n options. However, in big screen the menu doesn't located properly by default.
Screen Shot 2022-06-11 at 12 53 38

### Expected behavior
I think the menu should locate like this by default. (I'm now using tailwind class 'sm:-top-8' to fix this)
Screen Shot 2022-06-11 at 12 55 47

### Suggested solution

Maybe add some css like -top-8?

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) |    react-dropdown-menu     |   0.1.6      |
| React            | n/a     |     18.1.0    |
| Browser          |     Microsoft Edge |   100.0.1185.29      |
| Assistive tech   |     TailwindCSS    |     3.0.24    |
| npm/yarn         |     npm    |   8.6.0      |
| Operating System |     MacOS Monterey    |     12.3.1    |

### PR DESCRIPTION
**Pull request by benoitgrelard**
Supersedes #1516

Fixes #370 
Fixes #557
Fixes #740
Fixes #781 
Fixes #800
Fixes #1445
Fixes #1464

Others potentially linked to investigate:
#791 — explore this
#945 — may be doable now with use of mainAxis shift, but not sure it's worth it, probably needs a diff pattern
#1087 — linked to #791

---

> **Warning**
Docs updates:
- [ ] 🔥 `Arrow` `offset` prop is gone
- [ ] 🔥 `Content` `collisionTolerance` prop is renamed to `collisionPadding` and accept `number | Padding`
- [ ] ✨ `Content` `arrowPadding` prop is new
- [ ] ✨ `Content` `sticky` prop is new
- [ ] ✨ `Content` `hideWhenDetached` prop is new
- [ ] No need for `Portal` part around submenu contents (in `DropdownMenu` and `ContextMenu`)
- [ ] Update demos accordingly
---

- [ ] Also once PR is merged, go on `main` and delete references to `core/popper` in version files (only 1 I think)

### CODE REVIEW

File: packages/react/popper/src/Popper.tsx [hunk 1]
```diff
@@ -105,71 +126,136 @@ const PopperContent = React.forwardRef<PopperContentElement, PopperContentProps>
...
-      align,
-      alignOffset,
-      shouldAvoidCollisions: avoidCollisions,
-      collisionBoundariesRect,
-      collisionTolerance,
-    });
-    const isPlaced = placedSide !== undefined;
+    const isPlaced = x !== null && y !== null;
+    const [placedSide, placedAlign] = getSideAndAlignFromPlacement(placement);
+    const { arrow: { x: arrowX, y: arrowY } = {} } = middlewareData;
```
Comment (andy-hook) [inline packages/react/popper/src/Popper.tsx:178]:
```suggestion
    const arrowX = middlewareData.arrow?.x;
    const arrowY = middlewareData.arrow?.y;
```

Is this equivalent? I found the nested destructure trickier to follow at first glance.

File: packages/react/popper/src/Popper.tsx [hunk 2]
```diff
@@ -105,71 +126,136 @@ const PopperContent = React.forwardRef<PopperContentElement, PopperContentProps>
...
+      if (!isRoot) {
+        positionUpdateFns.add(update);
+        return () => {
+          positionUpdateFns.delete(update);
+        };
+      }
+    }, [isRoot, positionUpdateFns, update]);
+
+    // when nested contents are rendered in portals, they are appended out of order
+    // we need to re-compute the positioning once the parent has finally been placed.
```
Comment (andy-hook) [inline packages/react/popper/src/Popper.tsx:199]:
```suggestion
    // when nested contents are rendered in portals, they are appended out of order causing
    // children to be positioned incorrectly if initially open.
    // we need to re-compute the positioning once the parent has finally been placed.
    // https://github.com/floating-ui/floating-ui/issues/1531
```

How'd you feel about being really specific on this one?

File: packages/react/popper/src/Popper.tsx [hunk 3]
```diff
@@ -105,71 +126,136 @@ const PopperContent = React.forwardRef<PopperContentElement, PopperContentProps>
...
+      <div
+        ref={floating}
+        data-radix-popper-content-wrapper=""
+        style={{
+          position: strategy,
+          left: 0,
+          top: 0,
+          transform: isPlaced
+            ? `translate3d(${Math.round(x)}px, ${Math.round(y)}px, 0)`
+            : 'translate3d(0, -200vh, 0)',
```
Comment (andy-hook) [inline packages/react/popper/src/Popper.tsx:231]: I've forgotten the purpose of the `-200vh`, do we still need this?

File: packages/react/use-size/src/useSize.tsx [hunk 4]
```diff
@@ -30,10 +34,9 @@ function useSize(element: HTMLElement | SVGElement | null) {
           height = borderSize['blockSize'];
         } else {
           // for browsers that don't support `borderBoxSize`
-          // we calculate a rect ourselves to get the correct border box.
```
[the remark this answers is not in the document]
Comment (andy-hook) [reply packages/react/use-size/src/useSize.tsx:33]: Ah nice 👍  I vaguely remember running into something related to this on the nav menu.

File: packages/react/popper/src/Popper.tsx (hunk 3 above)
Comment (benoitgrelard) [reply packages/react/popper/src/Popper.tsx:231]: This is to make sure it's not overlapping other elements.
See here for more details on its origin: https://github.com/radix-ui/primitives/pull/541

**Review by andy-hook — APPROVED** [review]

**Pull request merged**